### PR TITLE
Make sluggable work with translatable slugs

### PR DESCRIPTION
**Pull request by mariavilaro**
Generate slugs for models that implement the Sluggable Trait and have translatable slugs

### CODE REVIEW

**Comment by mjauvin** [comment]
Can you describe the exact context where this can be used?

**Comment by mariavilaro** [comment]
If you have a model that implements the Sluggable Trait, and with a translatable slug:

    class User extends Model
    {
        use \October\Rain\Database\Traits\Sluggable;

        protected $slugs = ['slug' => 'name'];

        public $translatable = [ 'name', ['slug', 'index' => true] ];
}

When you save the model, only the slug for the default language is generated

**Comment by mjauvin** [comment]
Is there any issues this PR solves?

@w20k @bennothommo @osmanzeki can you test this and let me know if that works as expected?

**Comment by bennothommo** [comment]
@mariavilaro It might be better to make the new Sluggable trait extend the core `October\Rain\Database\Traits\Sluggable` class and adjust it to handle the translated slug properties as necessary, as there seems to be a lot of repeated code here.

It also should be an optional trait that can be used on translatable models, and not forced by the TranslatableModel behavior.

File: traits/Sluggable.php [hunk 1]
```diff
@@ -0,0 +1,113 @@
+<?php namespace RainLab\Translate\Traits;
+
+use October\Rain\Support\Str;
```
Comment (bennothommo) [inline traits/Sluggable.php]: @mariavilaro This can be shortened to just use the facade :)
```suggestion
use Str;
```

File: traits/Sluggable.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+trait Sluggable
+{
+    /**
+     * Adds translatable slug attributes to the translation data
+     * @return void
+     */
+    protected function slugAttributesTranslated($locale)
+    {
+        if (!isset($this->model->slugs) || !method_exists($this->model, 'slugAttributes')) return;
```
Comment (bennothommo) [inline traits/Sluggable.php]: @mariavilaro Please add braces around any condition blocks, as per [PSR-2](https://www.php-fig.org/psr/psr-2/) guidelines.
```suggestion
        if (!isset($this->model->slugs) || !method_exists($this->model, 'slugAttributes')) {
            return;
        }
```

File: traits/Sluggable.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        
+        $optionedAttributes = $this->getTranslatableAttributesWithOptions();
+
+        foreach ($this->model->slugs as $slugAttribute => $sourceAttributes) {
+            if (!isset($optionedAttributes[$slugAttribute])) continue;
```
Comment (bennothommo) [inline traits/Sluggable.php]: @mariavilaro Please add braces around any condition blocks, as per PSR-2 guidelines.

**Comment by mariavilaro** [comment]
> @mariavilaro It might be better to make the new Sluggable trait extend the core `October\Rain\Database\Traits\Sluggable` class and adjust it to handle the translated slug properties as necessary, as there seems to be a lot of repeated code here.

The core Sluggable class is already extended by the model, and their methods are already used by the default language slug generation, and the new trait also use some of them (e.j. $this->model->getSluggableSeparator()). I can't just override the methods, or then the core Sluggable class fails when the default language slug is generated. If there is a simpler way of doing this, I don't know how to do it.

> It also should be an optional trait that can be used on translatable models, and not forced by the TranslatableModel behavior.

If you want to use the Sluggable behavior, and also to have translated slugs, you're saying that you want your slugs to be generated, IMHO. The problem now is that the alternate language slugs are not automatically generated, and if you write them manually, they are not checked for uniqueness. Anyway, if you think that somebody could want to have translatable slugs that are not automatically populated or checked, maybe we could indicate that we want our slugs to behave as slugs by doing something like this:

    public $translatable = [ 'name', ['slug', 'index' => true, 'slug' => true] ];

But again, I think this is redundant.

**Comment by mariavilaro** [comment]
> Is there any issues this PR solves?

The problem now is that the alternate language slugs are not automatically generated, and if you write them manually, they are not checked for uniqueness

**Comment by bennothommo** [comment]
@mariavilaro Thanks for your response. There's a couple of things I'm suggesting we should avoid here:

- There's a reliance on the `October\Rain\Database\Traits\Sluggable` trait in your new Sluggable trait without it being explicitly defined as a dependency. If a developer who uses this new trait forgets to include the original trait, they may be confused as to why it is not working. This could be resolved by documentation  of course, but it would be cleaner if the new trait explicitly included the original trait in some way (eg. add a `use` declaration for the original Sluggable trait in the new Sluggable trait) and just use the new trait in any models that need translatable slug support.
- Given that both this new Sluggable trait and the old Sluggable trait share a lot of the same code but are completely independent, any future changes to the Sluggable trait are now possibly going to have to be made in two places, increasing the maintenance burden.
- Having it defined in the TranslatableModel behavior by default effectively means that additional functionality is being inserted into any model using the TranslatableModel behavior, even if they are not intending to use slugs. It would be better if this only occurs when needed.

Happy to hear your thoughts on the above.

**Comment by LukeTowers** [comment]
I agree with @bennothommo's comments

**Comment by LukeTowers** [comment]
To be clear, I would like to see an implementation closer to https://github.com/rainlab/translate-plugin/pull/478#issuecomment-489316751 before merging this.

**Comment by mariavilaro** [comment]
So @bennothommo you suggest to make a full working Sluggable trait, which extends Sluggable and which takes care of all slugs, both the default language and alternate languages.

So the model would use it like this:

    class User extends Model
    {
        use \RainLab\Translate\Traits\Sluggable;
        protected $slugs = ['slug' => 'name'];
        public $implement = ['RainLab.Translate.Behaviors.TranslatableModel'];
        public $translatable = [ 'name', ['slug', 'index' => true] ];
    }

Do I understand it right?

**Comment by bennothommo** [comment]
@mariavilaro That's correct. The new `RainLab\Translate\Traits\Sluggable` trait should be an extension of the functionality that already occurs with the original `October\Rain\Database\Traits\Sluggable` trait in which it handles the translated slugs, but also allows the original functionality to handle the default language.

The developer should only need to include this new trait if they want both the default and translated language slugs saved. If they want the original functionality where it only saves the default lanugage, they can continue to use the original trait.

**Comment by mjauvin** [comment]
@mariavilaro any progress on this?

**Comment by mjauvin** [comment]
@mariavilaro are you still interested in pursuing this?

**Comment by mjauvin** [comment]
Closing as it has been over a month since any activity on this occurred and we are trying to figure out what issues are still relevant. If this is still something that you would like to see through to fruition please respond and we can get the ball rolling.

**Comment by mariavilaro** [comment]
Hi sorry, I tried to make the independent RainLab\Translate\Traits\Sluggable and didn't succeed, I needed it to work urgently so I just made it work in my plugin. I don't really know how to make it work as a Translate Trait :(

**Comment by mariavilaro** [comment]
If someone wants to try it, this is the last try I did

```php
<?php namespace RainLab\Translate\Behaviors;

use System\Classes\ModelBehavior;
use Exception;
use Str;

class Sluggable extends ModelBehavior
{
    protected $translatedSlugsChanged = false;

    public function __construct($model)
    {
        $this->model = $model;

        $this->model->bindEvent('model.afterSave', function() {
            $this->afterSave();
        });
    }

    public function afterSave()
    {
        $locales = \RainLab\Translate\Models\Locale::listEnabled();
        $locales = array_keys($locales);
        $default_locale = \RainLab\Translate\Classes\Translator::instance()->getDefaultLocale();

        if (!method_exists($this->model, 'slugAttributes')) {
            throw new Exception(sprintf(
                'You must use the Sluggable trait in %s to use the Translate plugin Sluggable behaviour.', get_class($this->model)
            ));
        }

        if (!isset($this->model->slugs)) {
            throw new Exception(sprintf(
                'You must define a public $slugs property in %s to use the Translate plugin Sluggable behaviour.', get_class($this->model)
            ));
        }

        $optionedAttributes = $this->model->getTranslatableAttributesWithOptions();

        foreach ($this->model->slugs as $slugAttribute => $sourceAttributes) {
            if (!isset($optionedAttributes[$slugAttribute])) {
                throw new Exception(sprintf(
                    '$slugAttribute must be defined as an indexed translatable property in %s to use the Translate plugin Sluggable behaviour.', get_class($this->model)
                ));
            }
            foreach($locales as $locale) {
                if ($locale == $default_locale) {
                    continue;
                }
                $this->model->noFallbackLocale();
                $this->setSluggedValue($locale, $slugAttribute, $sourceAttributes);
            }
        }

        if ($this->translatedSlugsChanged) {
            //DOESN'T WORK
            //$this->model->save();
        }
    }

    protected function setSluggedValue($locale, $slugAttribute, $sourceAttributes, $maxLength = 175)
    {
        if (!$this->model->getAttributeTranslated($slugAttribute, $locale)) {
            if (!is_array($sourceAttributes)) {
                $sourceAttributes = [$sourceAttributes];
            }

            $slugArr = [];
            foreach ($sourceAttributes as $attribute) {
                $slugArr[] = $this->getSluggableSourceAttributeValue($locale, $attribute);
            }

            $slug = implode(' ', $slugArr);
            $slug = mb_substr($slug, 0, $maxLength);
            $slug = Str::slug($slug, $this->model->getSluggableSeparator());
        }
        else {
            $slug = $this->model->getAttributeTranslated($slugAttribute, $locale);
        }

        $unique_slug = $this->getSluggableUniqueAttributeValue($locale, $slugAttribute, $slug);

        if ($unique_slug != $this->model->getAttributeTranslated($slugAttribute, $locale)) {
            $this->translatedSlugsChanged = true;
            return $this->model->setAttributeTranslated($slugAttribute, $unique_slug, $locale);
        }

        return $unique_slug;
    }

    protected function getSluggableUniqueAttributeValue($locale, $name, $value)
    {
        $counter = 1;
        $separator = $this->model->getSluggableSeparator();
        $_value = $value;

        while (($this->model->methodExists('withTrashed') && $this->model->allowTrashedSlugs) ?
            $this->newSluggableQuery()->transWhere($name, $_value, $locale)->withTrashed()->count() > 0 :
            $this->newSluggableQuery()->transWhere($name, $_value, $locale)->count() > 0
        ) {
            $counter++;
            $_value = $value . $separator . $counter;
        }

        return $_value;
    }

    public function newSluggableQuery()
    {
        return $this->model->exists
            ? $this->model->newQuery()->where($this->model->getTable().'.'.$this->model->getKeyName(), '<>', $this->model->getKey())
            : $this->model->newQuery();
    }

    protected function getSluggableSourceAttributeValue($locale, $key)
    {
        if (strpos($key, '.') === false) {
            return $this->model->getAttributeTranslated($key, $locale);
        }

        $keyParts = explode('.', $key);
        $value = $this->model;
        foreach ($keyParts as $part) {
            if (!isset($value[$part])) {
                return null;
            }

            $value = $value[$part];
        }

        return $value;
    }
}
```

**Comment by bennothommo** [comment]
@mariavilaro Given that it was me who laid down the initial challenge, I'll be happy to give it a go. Thanks a lot for your effort so far.

**Comment by bennothommo** [comment]
Work continuing in #535.

**Pull request closed without merging**